File: src/PYB11/Solvers/OverlapNodeMap.py
from PYB11Generator import *
from MatrixMap import MatrixMap

@PYB11holder("std::shared_ptr")
@PYB11template("Dimension")
class OverlapNodeMap(MatrixMap):
    def pyinit(self,
               connectivity = "const FlatConnectivity<%(Dimension)s>&"):
        "Indexing for overlap connectivity"

    @PYB11virtual
    @PYB11const
    def firstGlobalIndex(self):
        "Global index of first node"
        return "int"

    @PYB11virtual
    @PYB11const
    def lastGlobalIndex(self):
        "Global index of last node"
        return "int"

    @PYB11virtual
    @PYB11const
    def numLocalElements(self):
        "Number of elements on this processor"
        return "int"

    @PYB11virtual
    @PYB11const
    def numGlobalElements(self):
        "Number of global elements"
        return "int"

    @PYB11virtual
    @PYB11const
    def numElementsPerRow(self,
                          localIndex = "int"):
        "Get the number of nonzero columns per row"
        return "int"
    
    @PYB11virtual
    @PYB11const
    def getColumnIndices(self,
                         localRowIndex = "int",
                         globalColumnIndices = "std::vector<int>&"):
        "Get the nonzero column indices for a given row"
        return "void"
    
    @PYB11virtual
    @PYB11const
    def checkClassInvariants(self):
        "Check class invariants"
        return "void"

File: src/Solvers/OverlapNodeMap.cc
//---------------------------------Spheral++----------------------------------//
// OverlapNodeMap
//
// Node map for overlap connectivity
//----------------------------------------------------------------------------//
#include "OverlapNodeMap.hh"

namespace Spheral {

//------------------------------------------------------------------------------
// Constructor
//------------------------------------------------------------------------------
template<typename Dimension>
OverlapNodeMap<Dimension>::
OverlapNodeMap(const FlatConnectivity<Dimension>& connectivity):
  NodeMap<Dimension>(connectivity) {
}

//------------------------------------------------------------------------------
// Get number of nonzero columns for this row
//------------------------------------------------------------------------------
template<typename Dimension>
int
OverlapNodeMap<Dimension>::
numElementsPerRow(int localRowIndex) const {
  return this->mConnectivity.numNonConstOverlapNeighbors(localRowIndex);
}

template<typename Dimension>
int
OverlapNodeMap<Dimension>::
numElementsPerRow(int nodeListIndex,
                  int nodeIndex) const {
  const auto localIndex = this->mConnectivity.nodeToLocal(nodeListIndex, nodeIndex);
  return numElementsPerRow(localIndex);
}

//------------------------------------------------------------------------------
// Get nonzero columns corresponding to this row
//------------------------------------------------------------------------------
template<typename Dimension>
void
OverlapNodeMap<Dimension>::
getColumnIndices(const int localRowIndex,
                 std::vector<int>& globalColumnIndices) const {
  this->mConnectivity.globalOverlapNeighborIndices(localRowIndex,
                                                   globalColumnIndices);
  return;
}

} // end namespace Spheral
